zerver/tests/test_service_bot_system.py

```diff
@@ -177,6 +179,88 @@ def test_entry_removal(self) -> None:
         self.assertTrue(storage.contains('another key'))
         self.assertRaises(StateError, lambda: storage.remove('some key'))
 
+    def test_internal_endpoint(self):
+        # type: () -> None
+        self.login(self.user_profile.email)
+        # Store some data.
+        initial_dict = {'key 1': 'value 1', 'key 2': 'value 2', 'key 3': 'value 3'}
+        params = {
+            'state': ujson.dumps(initial_dict)
+        }
+        result = self.client_put('/json/user_state', params)
+        self.assert_json_success(result)
+        # Assert the stored data for some keys.
+        params = {
+            'keys': ujson.dumps(['key 1', 'key 3'])
+        }
+        result = self.client_get('/json/user_state', params)
+        self.assert_json_success(result)
+        self.assertEqual(result.json()['state'], {'key 3': 'value 3', 'key 1': 'value 1'})
+        # Assert the stored data for all keys.
+        result = self.client_get('/json/user_state')
+        self.assert_json_success(result)
+        self.assertEqual(result.json()['state'], initial_dict)
+        # Store some more data; update an entry and store a new entry
+        dict_update = {'key 1': 'new value', 'key 4': 'value 4'}
+        params = {
+            'state': ujson.dumps(dict_update)
+        }
+        result = self.client_put('/json/user_state', params)
+        self.assert_json_success(result)
+        # Assert the data was updated.
+        updated_dict = initial_dict.copy()
+        updated_dict.update(dict_update)
+        result = self.client_get('/json/user_state')
+        self.assert_json_success(result)
+        self.assertEqual(result.json()['state'], updated_dict)
+        # Assert errors on invalid requests.
+        params = {  # type: ignore # Ignore 'incompatible type "str": "List[str]"; expected "str": "str"' for testing
+            'keys': ["This is a list, but should be a serialized string."]
+        }
+        result = self.client_get('/json/user_state', params)
+        self.assert_json_error(result, 'Argument "keys" is not valid JSON.')
+        params = {
+            'keys': ujson.dumps(["key 1", "nonexistent key"])
+        }
+        result = self.client_get('/json/user_state', params)
+        self.assert_json_error(result, "Cannot get state. <UserProfile: othello@zulip.com <Realm: zulip 1>> "
+                                       "doesn't have an entry with the key 'nonexistent key'.")
```

Comment:
So, the test suggests that we have a bug where we're printing a `UserProfile` in the JSON error.  We avoid doing that, since it's intended to be an internal implementation detail.  

I think we should be able to just give the complete error text as something clean, simple, and descriptive. like "Key does not exist.".